contracts/Pool.sol

```diff
@@ -1,31 +1,292 @@
+        uint32 weightCount;
+        // --- SLOT BOUNDARY
+        /// @notice The mapping of all the receivers to their weights, iterable
+        mapping(address => ReceiverWeight) receiverWeights;
+    }
+
+    struct Receiver {
+        /// @notice The block until which funds have been collected previously
+        uint64 lastCollectedCycle;
+        /// @notice The amount which has been previously collected
+        uint192 lastCollectedAmt;
+        // --- SLOT BOUNDARY
+        /// @notice The changes of collected amounts on specific cycle.
+        /// The keys are cycles, each cycle becomes collectable on block `C * cycleBlocks`
+        mapping(uint64 => int256) amtDeltas;
     }
 
+    /// @notice Details about all the senders, the key is the owner's address
     mapping(address => Sender) internal senders;
+    /// @notice Details about all the receivers, the key is the owner's address
+    mapping(address => Receiver) internal receivers;
+
+    /// @param _cycleBlocks The value of cycleBlocks to be used in the instance.
+    /// Low values make funds more available by shortening the average duration of funds being
+    /// frozen between being taken from senders' balances and being collectable by the receiver.
+    /// High values make collecting cheaper by making it process less cycles for a given time range.
+    constructor(uint64 _cycleBlocks) public {
+        cycleBlocks = _cycleBlocks;
+        /// @notice The block number of the last block which is described by the last cycle,
+        /// which is applied to receivers in block which fits in uint64
+        maxBlockNumber = (type(uint64).max / _cycleBlocks - 1) * _cycleBlocks;
+    }
+
+    /// @notice Returns amount of received funds available for collection
+    /// by the sender of the message
+    /// @return collected The available amount
+    function collectable() public view returns (uint256) {
+        Receiver storage receiver = receivers[msg.sender];
+        if (receiver.lastCollectedCycle == 0) return 0;
+        uint256 collected = 0;
+        int256 lastCollectedAmt = receiver.lastCollectedAmt;
+        uint64 lastCollectedCycle = receiver.lastCollectedCycle;
+        uint256 currFinishedCycle = block.number / cycleBlocks;
+        while (lastCollectedCycle < currFinishedCycle) {
+            lastCollectedCycle++;
+            lastCollectedAmt += receiver.amtDeltas[lastCollectedCycle];
+            collected += uint256(lastCollectedAmt);
+        }
+        return collected;
+    }
+
+    /// @notice Collects all received funds available for collection
+    /// by a sender of the message and sends them to that sender
+    function collect() public {
+        Receiver storage receiver = receivers[msg.sender];
+        if (receiver.lastCollectedCycle == 0) return;
+        uint64 lastCollectedCycle = receiver.lastCollectedCycle;
+        uint256 currFinishedCycle = block.number / cycleBlocks;
+        if (lastCollectedCycle == currFinishedCycle) return;
```

Comment:
I think it might be better to revert here and in other places where we simply return. The rationale is that instead the user probably wants to avoid sending transactions that are no-ops. It’s also likely that these no-op transaction constitute an error.

Reply:
From a user perspective it's perfectly fine to try to collect and end up with collecting 0 wei, because nobody sent me anything. There was no failure or mistake and the outcome is perfectly valid and expected, so in my opinion there's no need to raise an alarm. I'm not sure what is the convention in Ethereum, is a valid but no-op transaction considered an error?

Reply:
I see your point. I don’t know if there’s a convention around this. I’ll leave it up to you to decide

Reply:
I'm still convinced that the current solution is the right one